src/platforms/blackpill/platform.c

```diff
@@ -0,0 +1,144 @@
+#include "cdcacm.h"
+#include "usbuart.h"
+#include "morse.h"
+
+#include <libopencm3/stm32/rcc.h>
+#include <libopencm3/cm3/scb.h>
+#include <libopencm3/cm3/nvic.h>
+#include <libopencm3/stm32/exti.h>
+#include <libopencm3/stm32/usart.h>
+#include <libopencm3/stm32/syscfg.h>
+#include <libopencm3/usb/usbd.h>
+#include <libopencm3/cm3/systick.h>
+#include <libopencm3/cm3/cortex.h>
+#include <libopencm3/usb/dwc/otg_fs.h>
+
+
+jmp_buf fatal_error_jmpbuf;
+extern char _ebss[];
+
+void platform_init(void)
+{
+	volatile uint32_t *magic = (uint32_t *)_ebss;
+	/* Enable GPIO peripherals */
+	rcc_periph_clock_enable(RCC_GPIOA);
+	rcc_periph_clock_enable(RCC_GPIOC);
+	rcc_periph_clock_enable(RCC_GPIOB);
+
+	/* Check the USER button*/
+	if (gpio_get(GPIOA, GPIO0) ||
+		((magic[0] == BOOTMAGIC0) && (magic[1] == BOOTMAGIC1)))
+	{
+		magic[0] = 0;
+		magic[1] = 0;
+		/* Assert blue LED as indicator we are in the bootloader */
+		gpio_mode_setup(LED_PORT, GPIO_MODE_OUTPUT,
+						GPIO_PUPD_NONE, LED_BOOTLOADER);
+		gpio_set(LED_PORT, LED_BOOTLOADER);
+		/* Jump to the built in bootloader by mapping System flash.
+		   As we just come out of reset, no other deinit is needed!*/
+		rcc_periph_clock_enable(RCC_SYSCFG);
+		SYSCFG_MEMRM &= ~3;
+		SYSCFG_MEMRM |= 1;
+		scb_reset_core();
+	}
+	rcc_clock_setup_pll(&rcc_hse_25mhz_3v3[RCC_CLOCK_3V3_84MHZ]);
+
+	/* Enable peripherals */
+	rcc_periph_clock_enable(RCC_OTGFS);
+	rcc_periph_clock_enable(RCC_CRC);
+
+	/* Set up USB Pins and alternate function*/
+	gpio_mode_setup(GPIOA, GPIO_MODE_AF, GPIO_PUPD_NONE, GPIO9 | GPIO11 | GPIO12);
+	gpio_set_af(GPIOA, GPIO_AF10, GPIO9 | GPIO10 | GPIO11 | GPIO12);
+
+	GPIOA_OSPEEDR &= 0x3C00000C;
+	GPIOA_OSPEEDR |= 0x28000008;
+
+	gpio_mode_setup(JTAG_PORT, GPIO_MODE_OUTPUT,
+					GPIO_PUPD_NONE,
+					TCK_PIN | TDI_PIN);
+	gpio_mode_setup(JTAG_PORT, GPIO_MODE_INPUT,
+					GPIO_PUPD_NONE, TMS_PIN);
+	gpio_set_output_options(JTAG_PORT, GPIO_OTYPE_PP, GPIO_OSPEED_2MHZ,
+							TCK_PIN | TDI_PIN | TMS_PIN);
+	gpio_mode_setup(TDO_PORT, GPIO_MODE_INPUT,
+					GPIO_PUPD_NONE,
+					TDO_PIN);
+	gpio_set_output_options(TDO_PORT, GPIO_OTYPE_PP, GPIO_OSPEED_2MHZ,
+							TDO_PIN | TMS_PIN);
+
+	gpio_mode_setup(LED_PORT, GPIO_MODE_OUTPUT,
+					GPIO_PUPD_NONE,
+					LED_IDLE_RUN | LED_ERROR | LED_BOOTLOADER);
+
+	gpio_mode_setup(LED_PORT_UART, GPIO_MODE_OUTPUT, GPIO_PUPD_NONE, LED_UART);
+
+#ifdef PLATFORM_HAS_POWER_SWITCH
+	gpio_set(PWR_BR_PORT, PWR_BR_PIN);
+	gpio_mode_setup(PWR_BR_PORT, GPIO_MODE_OUTPUT,
+					GPIO_PUPD_NONE,
+					PWR_BR_PIN);
+#endif
+
+	platform_timing_init();
+	usbuart_init();
+	cdcacm_init();
+
+	// https://github.com/libopencm3/libopencm3/pull/1256#issuecomment-779424001
+	OTG_FS_GCCFG |= OTG_GCCFG_NOVBUSSENS | OTG_GCCFG_PWRDWN;
+	OTG_FS_GCCFG &= ~(OTG_GCCFG_VBUSBSEN | OTG_GCCFG_VBUSASEN);
+}
+
+void platform_srst_set_val(bool assert) { (void)assert; }
+bool platform_srst_get_val(void) { return false; }
+
+const char *platform_target_voltage(void)
+{
+	return NULL;
```

Comment:
This looks like it would be better implemented as returning a string like "Unknown"

Reply:
It seems that all other targets that don't have voltage measurement return NULL here. On some level it makes it easier to check if the platform has voltage support or not when calling this function instead of having to compare strings.

Reply:
Ah, hmm.. we'd seen something returning a string when it couldn't know voltage, but we buy your reasoning here.